source/common/common/inline_map.h

```diff
@@ -0,0 +1,460 @@
+#include <cstdint>
+#include <list>
+#include <memory>
+
+#include "envoy/common/optref.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/macros.h"
+#include "source/common/common/utility.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/node_hash_set.h"
+
+namespace Envoy {
+
+/**
+ * This file contains classes to help make very fast maps where a subset frequently used keys are
+ * known before the map is constructed.
+ *
+ * For example, the filter state always uses the filter name as the key and the filter name is
+ * known at compile time.
+ * By using the these classes, the filter state could get the key/value pair without any hash.
+ *
+ * Dynamic key lookups are also supported and have slightly worse performance than
+ * absl::flat_hash_map. Using this class makes sense if most of the lookups are expected to use the
+ * predefined handles.
+ *
+ *
+ * Example:
+ *
+ *   // Establishes a descriptor defining fast handles for the inline map. The descriptor must
+ *   // outlive maps defined using it. One possible usage model is to lazy-create statics for a
+ *   // descriptor and its handles using MUTABLE_CONSTRUCT_ON_FIRST_USE.
+ *   InlineMapDescriptor<std::string> descriptor;
+ *
+ *   // Create the handle for inline key. We should never do this after bootstrapping. Typically, we
+ *   // can do this by define a global variable.
+ *   InlineMapDescriptor<std::string>::Handle handle = descriptor.addHandle("inline_key");
+ *
+ *   // Finalize the descriptor. No further changes are allowed to the descriptor after this point.
+ *   descriptor.finalize();
+ *
+ *   // Create the inline map.
+ *   InlineMapPtr<std::string, std::string> inline_map = InlineMap<std::string,
+ *     std::string>::create(descriptor);
+ *
+ *   // Get value by handle.
+ *   inline_map->insert(handle, "value");
+ *   EXPECT_EQ(*inline_map->lookup(handle), "value");
+ */
+
+/**
+ * Maintains a collection of lightweight handles use for fast lookups into an inline map.
+ */
+template <class StorageKey> class InlineMapDescriptor {
+public:
+  /**
+   * Holds a fast index to a map. This is guaranteed to be small and cheap to copy and store.
+   */
+  class Handle {
+  public:
+    /**
+     * Get the id of the inline entry in the inline array. This could be used to access the
+     * key/value pair in the inline map without hash searching.
+     */
+    uint32_t inlineId() const { return inline_id_; }
+
+  private:
+    friend class InlineMapDescriptor;
+
+    // This constructor should only be called by InlineMapDescriptor.
+    Handle(uint32_t inline_id) : inline_id_(inline_id) {}
+
+    uint32_t inline_id_{};
+  };
+
+  using InlineKeysMap = absl::flat_hash_map<StorageKey, Handle>;
+  using InlineKeys = std::vector<StorageKey>;
+
+  InlineMapDescriptor() = default;
+
+  /**
+   * Add an inline key and return related handle. May only be called before finalize().
+   * Heterogeneous lookup is supported here.
```

Comment:
point out also that this allows sloppy init where if you add a key more than once you'll get the same Handle back.

Reply:
got it.

Reply:
does not look resolved?